### PR TITLE
feat(common-api): add HttpUrlFilter for host-based instrumentation filtering

### PR DESCRIPTION
**Pull request by benjoseph-grafana**
Adds a shared HttpUrlFilter contract that HTTP instrumentations can consult before starting an operation, so a filtered request produces no span, no HTTP client metrics, and no injected trace context.

Supports exact hosts and leading-wildcard subdomain patterns via onlyHosts() and exceptHosts(), with the functional interface as the escape hatch for arbitrary matching. Malformed patterns are rejected at construction rather than silently never matching at request time.

Not yet wired into any instrumentation.

See https://github.com/open-telemetry/opentelemetry-android/issues/1686#issuecomment-5688998606 for the proposal.

### CODE REVIEW

File: instrumentation/common-api/src/main/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilter.kt [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        fun validated(pattern: String): String {
+            val host = pattern.trim().lowercase()
+            require(host.isNotEmpty()) { "Host pattern must not be blank" }
+            require(host.none { it.isWhitespace() }) { "Host pattern '$pattern' must not contain whitespace" }
+            require(!host.contains('/') && !host.contains('?') && !host.contains('#')) {
+                "Host pattern '$pattern' must be a host, not a URL"
+            }
+            require(!host.contains(':') || host.startsWith('[')) {
+                "Host pattern '$pattern' must not include a port"
+            }
```
[the remark this answers is not in the document]
Comment (benjoseph-grafana) [reply instrumentation/common-api/src/main/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilter.kt]: @ is now rejected, so is IPv6 literals.

File: instrumentation/common-api/src/main/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilter.kt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        val host = hostOf(url) ?: return !instrumentOnMatch
+        return matches(host) == instrumentOnMatch
+    }
+
+    private fun matches(host: String): Boolean = host in exactHosts || subdomainSuffixes.any { host.endsWith(it) }
+
+    override fun toString(): String {
+        val name = if (instrumentOnMatch) "onlyHosts" else "exceptHosts"
+        return "HttpUrlFilter.$name(${(exactHosts + subdomainSuffixes.map { "*$it" }).joinToString()})"
+    }
```
[the remark this answers is not in the document]
Comment (benjoseph-grafana) [reply instrumentation/common-api/src/main/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilter.kt:109]: This is used only as a debugging aid and not a contract.

File: instrumentation/common-api/src/test/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilterTest.kt [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                "   " to "must not be blank",
+                "exa mple.com" to "whitespace",
+                "https://api.example.com" to "must be a host, not a URL",
+                "api.example.com/v1" to "must be a host, not a URL",
+                "api.example.com?q=1" to "must be a host, not a URL",
+                "api.example.com:8443" to "must not include a port",
+                "exa*ple.com" to "leading '*.' label",
+                "*" to "leading '*.' label",
+                "*." to "must name a domain after",
+            )
```
[the remark this answers is not in the document]
Comment (benjoseph-grafana) [reply instrumentation/common-api/src/test/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilterTest.kt:154]: done

File: instrumentation/common-api/src/test/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilterTest.kt [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+package io.opentelemetry.android.instrumentation.common.http
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import org.assertj.core.api.Assertions.assertThat
+import org.assertj.core.api.Assertions.assertThatThrownBy
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
```
[the remark this answers is not in the document]
Comment (benjoseph-grafana) [reply instrumentation/common-api/src/test/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilterTest.kt:14]: Follows the pattern in the repo, CI passes

File: instrumentation/common-api/src/main/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilter.kt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            // IPv6 literals are rejected as patterns, so such a URL has no host worth matching.
+            return if (host.startsWith('[')) null else host
+        }
+
+        /**
+         * Rejects patterns at configuration time rather than silently never matching at
+         * request time.
+         */
+        fun validated(pattern: String): String {
+            val host = pattern.trim().lowercase()
```
Comment (aranhave) [inline instrumentation/common-api/src/main/java/io/opentelemetry/android/instrumentation/common/http/HttpUrlFilter.kt:126]: Could we normalize Unicode host patterns to punycode, or reject them up front? `exceptHosts("bücher.example")` currently returns `true` for `https://xn--bcher-kva.example/`, which is how OkHttp represents the same host. A regression test for exact and wildcard patterns would help.

**Review by aranhave — COMMENTED** [review]
Thanks, Ben! The approach looks good. I left one question about a host-matching edge case to address before this lands.

**Comment by benjoseph-grafana** [comment]
Closing in favour of a different approach, after discussion at the Android SIG on
2026-09-17 (notes on #1686).

The short version: configuration moves to the agent DSL rather than a public
HttpUrlFilter interface, and filtering happens at export through a filtering
SpanExporter rather than inside each instrumentation. That means no new public
API and no instrumentation changes. Details and the per-component findings are
in https://github.com/open-telemetry/opentelemetry-android/issues/1686

@aranhave — your punycode catch carried over and is fixed in the new version.
Thanks for spotting it.

New PR to follow.
